File: app/integration_test/support/library_merge_seed.dart
import 'dart:io';

import 'package:sqlite3/sqlite3.dart';
import 'package:video_sniffing/engine/engine_host.dart';

const _sourceItemId = 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa';
const _targetItemId = 'bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb';
const _sourceEpId = 'cccccccc-cccc-cccc-cccc-cccccccccccc';
const _targetEpId = 'dddddddd-dddd-dddd-dddd-dddddddddddd';

/// 经 [EngineHost] 已打开的 data 目录写入两个同 title+season 的 Series 壳；
/// source 写入 idx=1，target 写入 idx=2（避免冲突），返回 (sourceItemId, targetItemId)。
Future<(String, String)> seedDuplicateSeriesForIntegration(
  EngineHost host,
  String dataDir, {
  String title = '示意剧',
  int season = 1,
}) async {
  final settings = host.settings();
  final mediaDir = Directory('$dataDir/${settings.mediaDir}');
  await mediaDir.create(recursive: true);

  final sourceFile = File('${mediaDir.path}/s1.mp4');
  final targetFile = File('${mediaDir.path}/t2.mp4');
  await sourceFile.writeAsBytes(const [0x78]);
  await targetFile.writeAsBytes(const [0x78]);

  final dbPath = '$dataDir/library.db';
  final db = sqlite3.open(dbPath);
  try {
    db.execute('BEGIN');
    db.execute('''
INSERT INTO library_items (id, kind, title, season, poster_path, created_at_ms) VALUES
  ('$_sourceItemId', 'series', '${_escapeSql(title)}', $season, NULL, 1),
  ('$_targetItemId', 'series', '${_escapeSql(title)}', $season, NULL, 2);
INSERT INTO library_episodes (id, item_id, idx, title, file_path, duration_ms, position_ms, source_url) VALUES
  ('$_sourceEpId', '$_sourceItemId', 1, '源1', '${_escapeSql(sourceFile.path)}', 10000, 0, NULL),
  ('$_targetEpId', '$_targetItemId', 2, '目标2', '${_escapeSql(targetFile.path)}', 10000, 0, NULL);
''');
    db.execute('COMMIT');
  } catch (error) {
    db.execute('ROLLBACK');
    rethrow;
  } finally {
    db.dispose();
  }

  return (_sourceItemId, _targetItemId);
}

String _escapeSql(String value) => value.replaceAll("'", "''");

File: engine/tests/support/library_merge_seed.rs
use std::fs;
use uuid::Uuid;
use video_sniffing_engine::test_api::LibraryStore;
use video_sniffing_engine::{Engine, LibraryEpisode, LibraryItem, LibraryItemKind};

pub struct DuplicateSeriesSeed {
    pub source_item_id: String,
    pub target_item_id: String,
}

pub fn seed_duplicate_series(
    engine: &Engine,
    title: &str,
    season: Option<u32>,
) -> DuplicateSeriesSeed {
    let media_dir = engine.media_dir();
    let data_dir = media_dir.parent().unwrap();
    let store = LibraryStore::open(&data_dir.join("library.db")).unwrap();
    let source_item_id = Uuid::new_v4().to_string();
    let target_item_id = Uuid::new_v4().to_string();
    store
        .upsert_item(&LibraryItem {
            id: source_item_id.clone(),
            kind: LibraryItemKind::Series,
            title: title.into(),
            season,
            poster_path: None,
            created_at_ms: 1,
        })
        .unwrap();
    store
        .upsert_item(&LibraryItem {
            id: target_item_id.clone(),
            kind: LibraryItemKind::Series,
            title: title.into(),
            season,
            poster_path: None,
            created_at_ms: 2,
        })
        .unwrap();
    DuplicateSeriesSeed {
        source_item_id,
        target_item_id,
    }
}

pub fn add_episode(
    engine: &Engine,
    item_id: &str,
    index: u32,
    title: &str,
    file_name: &str,
    position_ms: i64,
) -> LibraryEpisode {
    let media = engine.media_dir().join(file_name);
    fs::write(&media, b"x").unwrap();
    let store = LibraryStore::open(
        &engine
            .media_dir()
            .parent()
            .unwrap()
            .join("library.db"),
    )
    .unwrap();
    let ep = LibraryEpisode {
        id: Uuid::new_v4().to_string(),
        item_id: item_id.into(),
        index,
        title: title.into(),
        file_path: media.to_string_lossy().into(),
        duration_ms: Some(10_000),
        position_ms,
        source_url: None,
    };
    store.upsert_episode(&ep).unwrap();
    ep
}
